platform/api/esx/api.go

```diff
@@ -0,0 +1,742 @@
+		return nil, nil, err
+	}
+	if cisr.Error != nil {
+		return nil, nil, errors.New(cisr.Error[0].LocalizedMessage)
+	}
+	return archive, cisr, nil
+}
+
+func (a *API) getServerDefaults() (ServerResources, error) {
+	finder := find.NewFinder(a.client.Client, true)
+	datacenter, err := finder.DefaultDatacenter(a.ctx)
+	if err != nil {
+		return ServerResources{}, err
+	}
+	finder.SetDatacenter(datacenter)
+	resourcePool, err := finder.DefaultResourcePool(a.ctx)
+	if err != nil {
+		return ServerResources{}, err
+	}
+	datastore, err := finder.DefaultDatastore(a.ctx)
+	if err != nil {
+		return ServerResources{}, err
+	}
+
+	defaultNetwork, err := finder.DefaultNetwork(a.ctx)
+	if err != nil {
+		return ServerResources{}, err
+	}
+
+	return ServerResources{
+		finder:       finder,
+		datacenter:   datacenter,
+		resourcePool: resourcePool,
+		datastore:    datastore,
+		network:      defaultNetwork,
+	}, nil
+}
+
+func (a *API) uploadToResourcePool(archive *Archive, resourcePool *object.ResourcePool, cisr *types.OvfCreateImportSpecResult, folder *object.Folder) (*types.ManagedObjectReference, error) {
+	lease, err := resourcePool.ImportVApp(a.ctx, cisr.ImportSpec, folder, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := lease.Wait(a.ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var items []ovfFileItem
+
+	for _, device := range info.DeviceUrl {
+		for _, item := range cisr.FileItem {
+			if device.ImportKey != item.DeviceId {
+				continue
+			}
+
+			u, err := a.client.Client.ParseURL(device.Url)
+			if err != nil {
+				return nil, err
+			}
+
+			i := ovfFileItem{
+				url:  u,
+				item: item,
+				ch:   make(chan progress.Report),
+			}
+
+			items = append(items, i)
+		}
+	}
+
+	upd := newLeaseUpdater(a.client.Client, lease, items)
+	defer upd.Done()
+
+	for _, i := range items {
+		err = a.upload(archive, lease, i)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = lease.HttpNfcLeaseComplete(a.ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &info.Entity, nil
+}
+
+func networkMap(finder *find.Finder, e *ovf.Envelope) (p []types.OvfNetworkMapping) {
+	networks := map[string]string{}
+
+	if e.Network != nil {
+		for _, net := range e.Network.Networks {
+			networks[net.Name] = net.Name
+		}
+	}
+
+	for src, dst := range networks {
+		if net, err := finder.Network(context.TODO(), dst); err == nil {
+			p = append(p, types.OvfNetworkMapping{
+				Name:    src,
+				Network: net.Reference(),
+			})
+		}
+	}
+	return
+}
+
+func (a *API) updateOVFEnv(vm *object.VirtualMachine, ignitionData string) error {
+	var property []types.VAppPropertySpec
+
+	var mvm mo.VirtualMachine
+	err := vm.Properties(a.ctx, vm.Reference(), []string{"config", "config.vAppConfig", "config.vAppConfig.property"}, &mvm)
+	if err != nil {
+		return fmt.Errorf("couldn't get config.vappconfig: %v", err)
+	}
+
+	for _, item := range mvm.Config.VAppConfig.(*types.VmConfigInfo).Property {
+		if item.Id == "guestinfo.coreos.config.data" {
```

Comment:
Does coreos-cloudinit also support this property?  If so, we should call it `userdata` rather than `ignitionData`.  If not, the top-level API call should error out if passed a `conf` which is not `IsIgnition`.